### PR TITLE
dot/core, lib/grandpa, lib/scale: implement FinalityGadget interface, implement ability to convert grandpa messages to network messages

### LINKED ISSUE #933: grandpa: convert VoteMessage, Justification, and FinalizationMessage to network messages
## Expected Behavior

- add VoteMessage, Justification, and FinalizationMessage to network messages
- have them implement the `network.Message` interface
- update grandpa channels to have type `network.Message`

## Current Behavior

- grandpa messages are in grandpa package
- `in`, `out` chans only use `VoteMessage`

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [ ] I am planning to submit a pull request to fix this issue myself

### LINKED ISSUE #934: grandpa: broadcast FinalizationMessage upon block finalization
## Expected Behavior

- see `grandpa.playGrandpaRound` and `grandpa.attemptToFinalize`
- requires #933 
- also need to implement justification collecting

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [x] I am planning to submit a pull request to fix this issue myself

### PR DESCRIPTION
**Pull request by noot**
## Changes

- implement core.FinalityGadget and core.FinalityMessage interfaces
- update grandpa service to fulfil core.FinalityGadget
- update grandpa VoteMessage and FinalizationMessage to implement core.FinalityMessage

## Tests

```
go test ./... -short
```

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [x] I have reviewed my own pull request before requesting a review
- [x] All integration tests and required coverage checks are passing

## Issues

- closes #933 closes #934

### CODE REVIEW

File: lib/grandpa/message.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+// GetFinalizedChannel returns a read-only FinalizationMessage channel
+func (s *Service) GetFinalizedChannel() <-chan FinalityMessage {
+	return s.finalized
+}
+
+// DecodeMessage decodes a network-level consensus message into a GRANDPA VoteMessage or FinalizationMessage
+func (s *Service) DecodeMessage(msg *ConsensusMessage) (FinalityMessage, error) {
+	m, err := scale.Decode(msg.Data, &VoteMessage{Message: new(SignedMessage)})
+	if err != nil {
+		// try FinalizatioNmessage
```
Comment (ryanchristo) [inline lib/grandpa/message.go:39]: typo

File: lib/common/common.go [hunk 2]
```diff
@@ -78,6 +78,31 @@ func HexToBytes(in string) ([]byte, error) {
 	return out, err
 }
 
+// MustHexToBytes turns a 0x prefixed hex string into a byte slice
+// it panic if it cannot decode the string
+func MustHexToBytes(in string) []byte {
+	if len(in) < 2 {
```
Comment (edwardmack) [inline lib/common/common.go:84]: I like this, should we replace usages of `HexToBytes` with this?  Or is there use for `HexToBytes`?

File: lib/common/hash.go [hunk 3]
```diff
@@ -101,3 +101,21 @@ func HexToHash(in string) (Hash, error) {
 	copy(buf[:], out)
 	return buf, err
 }
+
+// MustHexToHash turns a 0x prefixed hex string into type Hash
+// it panics if it cannot turn the string into a Hash
+func MustHexToHash(in string) Hash {
+	if strings.Compare(in[:2], "0x") != 0 {
```
Comment (edwardmack) [inline lib/common/hash.go:108]: Same question, do we need `HexToHash` which returns error, or this which may panic?

File: lib/grandpa/grandpa.go [hunk 4]
```diff
@@ -232,7 +260,7 @@ func (s *Service) playGrandpaRound() error {
 				log.Trace("[grandpa] failed to check if round is completable", "error", err)
 			}
 
-			finalizable, err := s.isFinalizable(s.state.round - 1)
+			finalizable, err := s.isFinalizable(s.state.round)
```
Comment (edwardmack) [inline lib/grandpa/grandpa.go:263]: Used to check s.state.round -1, now checking current round.  I don't understand why?

Comment (noot) [reply lib/grandpa/grandpa.go:265]: if the current round isn't finalizable and the `receiveMessages` function returns, then we won't be able to receive any more votes and the `attemptToFinalizeRound` gets stuck in an infinite loop, I was seeing this happen pretty often, so I changed it

File: lib/common/common.go (hunk 2 above)
Comment (noot) [reply lib/common/common.go:84]: I think usages in the tests could be replaced, but in the actual code let's leave it as is, so it doesn't panic

File: lib/common/hash.go (hunk 3 above)
Comment (noot) [reply lib/common/hash.go:108]: same as above, it's nice to have a version that doesn't panic

**Comment by noot** [comment]
@noot add scale unit tests

**Pull request merged**